Extension metadata: detecting the platform

carlopi opened the pull request:
### The plan
First step towards extensions metadata.

I separated them to make it somehow easier to follow the logic / test them in isolation.

This PR has as main goal of making the duckdb_platform to be stored within extension available in the cmake environment.

### Detecting / providing the platform
This is done by compiling a free-standing program (based on the logic already present within duckdb) that once compiled and run produces as ouptut the platform reported by `PRAGMA platform;`.
This executable is then run, it's output saved as content of a file in the cmake build directory.

In case we are cross compiling, it's always possible to provide the platform externally by passing the make variable DUCKDB_PLATFORM.

Example, after `GEN=ninja make`:
```
$ ./build/release/duckdb -c "PRAGMA platform;"
┌───────────┐
│ platform  │
│  varchar  │
├───────────┤
│ osx_arm64 │
└───────────┘
$ ./build/release/duckdb_platform_binary
osx_arm64
$ cat build/release/duckdb_platform_out
osx_arm64
```
while after `DUCKDB_PLATFORM=unknown GEN=ninja make`
```
$ ./build/release/duckdb -c "PRAGMA platform;"
┌───────────┐
│ platform  │
│  varchar  │
├───────────┤
│ osx_arm64 │
└───────────┘
$ ./build/release/duckdb_platform_binary  
zsh: no such file or directory: ./build/release/duckdb_platform_binary
$ cat build/release/duckdb_platform_out
unknown
```

Basically the DuckDB platform used for the current binary will be still be decided with the current logic, but the secondary DuckDB platform (either autodetected or provided) will be saved to the duckdb_platform_out file. This secondary platform is meant to be used for building, singing and deploying extensions, and it can be seen on the touched workflow files the `DUCKDB_PLATFORM` is basically a stand in for the `deploy_as` YAML variable.

### Using the platform, part 1
As a bonus /quality of life improvement, after this PR extensions are also copied over to a folder structure compatible with the `SET extension_directory = 'build/release/repository'` directory.

Example, after building with
`EXTENSION_CONFIGS=.github/config/bundled_extensions.cmake GEN=ninja make`
the extensions will be copied into the right folder structure like:
```
carlo@ScroogeMcDuck duckdb % tree build/release/repository
build/release/repository
└── 988139051e
    └── osx_arm64
        ├── autocomplete.duckdb_extension
        ├── excel.duckdb_extension
        ├── fts.duckdb_extension
        ├── icu.duckdb_extension
        ├── inet.duckdb_extension
        ├── json.duckdb_extension
        ├── parquet.duckdb_extension
        ├── sqlsmith.duckdb_extension
        ├── tpcds.duckdb_extension
        ├── tpch.duckdb_extension
        └── visualizer.duckdb_extension

3 directories, 11 files
```
that allows for example:
```
carlo@ScroogeMcDuck duckdb % ./build/release/duckdb -unsigned -cmd "SET extension_directory='build/release/repository';" 
v0.9.3-dev3313 988139051e
Enter ".help" for usage hints.
D FROM duckdb_extensions() WHERE extension_name = 'tpcds';
┌────────────────┬─────────┬───────────┬─────────────────────────────────────────────────────────────────────────────┬──────────────────────────────────────────────┬───────────┐
│ extension_name │ loaded  │ installed │                                install_path                                 │                 description                  │  aliases  │
│    varchar     │ boolean │  boolean  │                                   varchar                                   │                   varchar                    │ varchar[] │
├────────────────┼─────────┼───────────┼─────────────────────────────────────────────────────────────────────────────┼──────────────────────────────────────────────┼───────────┤
│ tpcds          │ false   │ true      │ build/release/repository/988139051e/osx_arm64/tpcds.duckdb_extension │ Adds TPC-DS data generation and query support │ []        │
└────────────────┴─────────┴───────────┴─────────────────────────────────────────────────────────────────────────────┴──────────────────────────────────────────────┴───────────┘
D CALL dsdgen(sf = 1);
Error: Catalog Error: Table Function with name "dsdgen" is not in the catalog, but it exists in the tpcds extension.

Please try installing and loading the tpcds extension by running:
INSTALL tpcds;
LOAD tpcds;

Alternatively, consider enabling auto-install and auto-load by running:
SET autoinstall_known_extensions=1;
SET autoload_known_extensions=1;
D LOAD tpcds;
D CALL dsdgen(sf = 1);
```

Possibly a short-hand for `./build/release/duckdb -unsigned -cmd "SET extension_directory='build/release/repository';"` can be handy, and possibly the same logic can be used within the unit-tester to more expand the testing situations, but those two expansions are to be built on top.


---
Edited with latest changes

carlopi committed:
Move DuckDB::Platform() to header

carlopi committed:
Infrastructure: Add utility to return duckdb_platform

carlopi committed:
platform: Log and save result to CMAKE variable

carlopi committed:
Default LOCAL_EXTENSION_REPO to  ${CMAKE_BINARY_DIR}/deploy

carlopi committed:
Infra: Compute extension folder name in CMake

carlopi committed:
Rework local_extension_repo logic to be done via CMake

Now after having built, say by GEN=ninja make
a compatible extension repository will be populated in the
./build/release/.duckdb/extensions/ folder

And launching the compiled duckdb cli program with:
./build/release/duckdb -unsigned -cmd "SET extension_directory='build/release'"
allows to load the just built extensions.

carlopi committed:
DUCKDB_PLATFORM either empty or non defined implies autodetection

carlopi committed:
Pass through duckdb_arch in the yaml files in case of cross-compilation

In the other cases autodetection should work

carlopi committed:
Pass platfrom also from _extension_distribution

carlopi committed:
More fixes, also handle duckdb-wasm

carlopi committed:
Move implementation up while changing where dependecies are inserted

carlopi committed:
Various fixes to make it work on all platforms

carlopi committed:
Crosscompiling for Java

carlopi committed:
Fix test loadable extensions to follow extension/name/name.duckdb_extension pattern

format

carlopi committed:
Add DUCKDB_PLATFORM to OSX.yml

carlopi committed:
Add DUCKDB_PLATFORM to LinuxRelease.yml

carlopi committed:
cmake: Fix Windows copying of extensions with globs

carlopi committed:
Revert "Fix test loadable extensions to follow extension/name/name.duckdb_extension pattern"

This reverts commit fe379c722ceb855b47d9bd1466df8f725056295b.

samansmink commented on .github/actions/build_extensions/action.yml:
```diff
@@ -27,6 +27,9 @@ inputs:
     default: ''
 
   # Build config
+  duckdb_arch:
```
could this be the same param as `deploy_as`? It should always be the same value, I think?

carlopi replied:
Yes and no, they should always the same if they are both provided, but there can be cases where either `duckdb_arch` is not provided since it can be autodetected (say we are not cross-compiling) or cases where `deploy_as` is not provided since it doubles as "should it be also deployed".

A next step would be swapping `deploy_as` with a `also_deploy` boolean (that default to false) to avoid duplications / risk of them drifting apart, but I would prefer first to land it as is and double check stuff is working properly.

samansmink commented on .github/workflows/Java.yml:
```diff
@@ -98,6 +98,7 @@ jobs:
       ENABLE_EXTENSION_AUTOINSTALL: 1
       OVERRIDE_JDBC_OS_ARCH: arm64
       JAVA_HOME: ../../jdk8u345-b01
+      DUCKDB_PLATFORM: linux_amd64
```
shouldn't this be `arm64`?

carlopi replied:
+1

samansmink commented on CMakeLists.txt:
```diff
@@ -636,8 +650,30 @@ function(link_threads LIBRARY)
 
+# Deploys extensions to a local repository (a folder structure that contains the duckdb version + binary arch)
+if ("${LOCAL_EXTENSION_REPO}" STREQUAL "")
+  set(LOCAL_EXTENSION_REPO ${CMAKE_BINARY_DIR}/.duckdb/extensions)
```
maybe `${CMAKE_BINARY_DIR}/.duckdb/extensions` is not the most intuitive path? I would prefer something like `${CMAKE_BINARY_DIR}/extension_deploy` or `${CMAKE_BINARY_DIR}/.duckdb/extension_repository`

no super strong opinion here, feel free to ignore

carlopi replied:
Problem here is that:
```
D SET extension_directory = 'path/to/some/folder';
D LOAD x;
Error: IO Error: Extension "path/to/some/folder/.duckdb/extensions/fe379c722c/osx_arm64/x.duckdb_extension" not found.
```
where the `.duckdb/extensions` is added by default.

---

My preferred option here would be changing the semantic of extension_directory, where if not specified it defaults to `fs.GetHomeDirectory() + ".duckdb/extension"`, so basically it becomes the base folder of the extension hierarchy.

This will probably break someone's workflows (eg https://github.com/search?q=duckdb+%22SET+extension_directory%22&type=code&p=1), so might need some care, potentially changing name to `extension_base_directory` so an error will be thrown at least.

On the other side, better break stuff today than in a few months...
On this I am not sure, I would push this PR in and in parallel we can set on this semantic (given it's only using extension_directory default, I would argue it's somehow independent).

samansmink reviewed (commented):
Looks great! Nice that things are now handled by cmake as opposed to my ugly script. Added a few small comments, but the rest LGTM!

carlopi committed:
Addressing @samansmink's review

carlopi committed:
Bring back python script, and make python3 compulsory when LOCAL_EXTENSION_REPO is set

carlopi committed:
Include MACRO definition for duckdb-wasm

carlopi commented:
Changed course and went back to using a python script to populate the extension directory.

I had not found a proper way to go around cmake on various platforms that was not error prone / confusing (to me and I guess to whoever would come next).

When providing a LOCAL_EXTENSION_REPO folder, I added an explicit dependency on python3 using the `find_package (Python3)` functionality + testing against the `Python3_FOUND` variable.

carlopi committed:
create_local_extension_repo: fix Windows handling

carlopi committed:
Avoid find_package(python3) and packaging extension with CLANG_TIDY

Unsure why this is actually needed, but...

carlopi committed:
Add back support for duckdb-wasm extensions

carlopi commented:
Latest problem was that somehow `make check-tidy` is incompatible with `find_package(Python3)` in the sense that it shows a bunch of errors in the pythonpkg. This might be worth investigating since there might be some assumption about Python headers or other problems, but it's not really connected to this PR and I would be for moving forward.

carlopi committed:
Fix error in passing through variable

samansmink reviewed (approved):
LGTM!

Mytherin commented on CMakeLists.txt:
```diff
@@ -1084,6 +1130,35 @@ if(BUILD_PYTHON)
+  set_target_properties(duckdb_platform_binary PROPERTIES RUNTIME_OUTPUT_DIRECTORY
+                                         ${PROJECT_BINARY_DIR})
+  if (WIN32)
+    add_custom_target(
```
Maybe I'm missing something but this if/else seems to not change anything?

carlopi replied:
Right, nice catch. This has been refactored away, now using properly CMake to detect the binary (instead of building the name on the spot). I will take this out.

carlopi replied:
Handled in follow up commit, thanks

Mytherin reviewed (commented):
Thanks! LGTM - one minor comment

carlopi committed:
CMake: Avoid unnecessary branch (review from @Mytherin)

carlopi commented:
Fixed the unnecessary branch, for me this can go in

Mytherin commented:
Thanks!
